File: tree/Traversal/Inorder/iterative/python/InorderTraversal.py
from collections.abc import Sequence


def InorderTraversal(left_children: Sequence[int], right_children: Sequence[int], root: int) -> list[int]:
    order: list[int] = []
    stack: list[int] = []
    current = root

    while current != -1 or stack:
        while current != -1:
            stack.append(current)
            current = left_children[current]

        current = stack.pop()
        order.append(current)
        current = right_children[current]

    return order

File: tree/Traversal/Inorder/iterative/rust/src/InorderTraversal.rs
#[allow(non_snake_case)]
pub fn InorderTraversal(left_children: &[i32], right_children: &[i32], root: usize) -> Vec<usize> {
    let mut order = Vec::new();
    let mut stack = Vec::new();
    let mut current = Some(root);

    while current.is_some() || !stack.is_empty() {
        while let Some(node) = current {
            stack.push(node);
            let left_child = left_children[node];
            current = if left_child == -1 {
                None
            } else {
                Some(left_child as usize)
            };
        }

        let node = stack.pop().unwrap();
        order.push(node);
        let right_child = right_children[node];
        current = if right_child == -1 {
            None
        } else {
            Some(right_child as usize)
        };
    }

    order
}
